### PR TITLE
[PT FE] Support scalar (rank-0) input for aten::log_softmax and aten::softmax

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
PyTorch allows log_softmax and softmax on scalar (rank-0) tensors by internally
reshaping them to 1D. The WeNet decoder model hits this path — torch.tensor(0.0)
is passed through log_softmax with dim=-1, which fails during conversion because
OpenVINO core ops reject axis=-1 for rank-0 inputs.

This patch adds scalar handling at multiple levels:

- PT frontend translators: unsqueeze scalar to rank 1, apply the op with axis=0,
  squeeze back. This mirrors what PyTorch does internally.
- Core op validation (LogSoftmax v5, Softmax v1/v8): skip axis bounds check for
  rank-0, since the result is mathematically determined (softmax=1, log_softmax=0).
- Softmax v8 evaluate: treat rank-0 as a 1-element tensor.
- Decomposition passes (LogSoftmax, Softmax): replace rank-0 input with the
  appropriate constant (0 or 1).
- Softmax v8-to-v1 downgrade: replace rank-0 with constant 1 instead of
  attempting axis normalization.

The core op and transformation changes are necessary even though the PT frontend
translator already reshapes scalars to 1D. The reason is that OpenVINO's
validation and transformation pipeline processes operations at multiple stages,
and rank-0 LogSoftmax/Softmax nodes can appear in the graph from paths other than
the PT frontend — for example, from ONNX frontend, from direct graph construction
via the C++ API, or from subgraph extraction during testing. Additionally, the
reference implementations (evaluate methods) are used by the constant folding pass
and by the test infrastructure to compute expected outputs. If a rank-0
LogSoftmax/Softmax node reaches constant folding or the interpreter backend
without these fixes, it will hit the same axis validation failure. The
decomposition passes need the same treatment because they run before plugin-
specific lowering and would otherwise crash on the axis normalization step when
encountering a rank-0 input. In short, every layer that touches the axis attribute
of these ops needs to gracefully handle the rank-0 case to avoid breaking
whichever path happens to encounter it first.

### Tickets:
 - 143262

### CODE REVIEW

File: src/core/src/op/log_softmax.cpp [hunk 1]
```diff
@@ -21,14 +21,19 @@ bool op::v5::LogSoftmax::visit_attributes(AttributeVisitor& visitor) {
...
-                              m_axis < input_shape.rank().get_length() && m_axis >= -input_shape.rank().get_length(),
-                              "Reduction axis (",
-                              m_axis,
-                              ") is out of bounds (argument shape: ",
-                              input_shape,
-                              ").");
+    if (input_shape.rank().is_static()) {
+        const auto rank = input_shape.rank().get_length();
+        // Rank-0 (scalar) inputs are valid: log_softmax of a single value is always 0.
+        if (rank > 0) {
```
Comment (praasz) [inline src/core/src/op/log_softmax.cpp]:
```suggestion
    if (const auto& rank = input_shape.rank(); rank.is_static()) {
        ov::util::validate_axis(m_axis, rank, *this)
```
or use `ov::util::is_axis_valid` which should include axis==0 check (ignore rank)

File: src/core/src/op/softmax.cpp [hunk 2]
```diff
@@ -42,14 +42,19 @@ bool Softmax::visit_attributes(AttributeVisitor& visitor) {
...
     const auto& input_shape = get_input_partial_shape(0);
-    if (input_shape.rank().is_static())
-        NODE_VALIDATION_CHECK(this,
-                              m_axis < static_cast<size_t>(input_shape.rank().get_length()),
-                              "Reduction axis (",
-                              m_axis,
-                              ") is out of bounds (argument shape: ",
-                              input_shape,
-                              ").");
+    if (input_shape.rank().is_static()) {
```
Comment (praasz) [inline src/core/src/op/softmax.cpp]: See comment of  log_softmax

File: src/core/src/op/softmax.cpp [hunk 3]
```diff
@@ -136,6 +144,24 @@ bool Softmax::evaluate(TensorVector& outputs, const TensorVector& inputs) const
 
     const auto& input_shape = inputs[0].get_shape();
     const auto rank = static_cast<int64_t>(input_shape.size());
+    outputs[0].set_shape(input_shape);
+
+    // Rank-0 (scalar): treat as 1-element 1D tensor; softmax of a single value is always 1.
+    if (rank == 0) {
+        using namespace ov::element;
+        return IF_TYPE_OF_CONVERT_TENSORS(v8_Softmax_evaluate,
```
Comment (praasz) [inline src/core/src/op/softmax.cpp]: This will increase binary size, re-use common entry for caculations

File: src/core/src/op/softmax.cpp [hunk 4]
```diff
@@ -136,6 +144,24 @@ bool Softmax::evaluate(TensorVector& outputs, const TensorVector& inputs) const
...
+                                          OV_PP_ET_LIST(f32, f64),
+                                          softmax::Evaluate,
+                                          inputs[0].get_element_type(),
+                                          inputs[0],
+                                          outputs[0],
+                                          Shape{1},
+                                          AxisSet{0});
+    }
+
     OPENVINO_ASSERT(-rank <= m_axis && m_axis < rank,
```
Comment (praasz) [inline src/core/src/op/softmax.cpp]: If some common util used this could be removed check `validation_util.hpp` for axis normalization utils

File: src/core/src/op/log_softmax.cpp (hunk 1 above)
Comment (evkotov) [reply src/core/src/op/log_softmax.cpp]: Thanks for the suggestion! Switched to ov::util::validate_axis for both LogSoftmax v5 and Softmax v8.

However, I had to keep a rank > 0 guard:


if (const auto& rank = input_shape.rank(); rank.is_static() && rank.get_length() > 0) {
    ov::util::validate_axis(m_axis, rank, *this);
}
The reason: is_axis_valid(-1, 0) returns false (for rank-0 only axis=0 is accepted), but the PT frontend test framework calls validate_nodes_and_infer_types() after re-assigning parameter shapes from actual input data. When a scalar input is used with dim=-1 (standard PyTorch convention), the LogSoftmax/Softmax node in the graph still has axis=-1 — the unsqueeze/squeeze wrapper is around it, but the node itself stores the original axis. After shape re-inference with rank-0 partial shape, validate_axis(-1, rank=0) would fail.

Skipping validation for rank-0 is safe since the result is mathematically determined (softmax=1, log_softmax=0) regardless of axis value.

For v1::Softmax I kept NODE_VALIDATION_CHECK with r == 0 || m_axis < r because m_axis is size_t — casting it to int64_t for validate_axis would turn (size_t)-1 into -1 which is_axis_valid would accept for rank >= 1.

Also eliminated the duplicated IF_TYPE_OF_CONVERT_TENSORS block in v8::Softmax::evaluate() as requested — now uses a single code path with normalized shape/axis.

Comment (praasz) [reply src/core/src/op/log_softmax.cpp]: The frontend must apply correct conversion the rank = 0 and dim = -1 is not correct
cannot get axis from back if there is no axis

Comment (evkotov) [reply src/core/src/op/log_softmax.cpp]: fixed

File: src/core/src/validation_util.cpp [hunk 5]
```diff
@@ -324,7 +324,8 @@ bool has_no_symbols(const ov::TensorSymbol& symbols) {
 }
 
 bool is_axis_valid(int64_t axis, int64_t rank) {
-    return (axis == 0) || (-rank <= axis && axis < rank);
+    // For rank-0 (scalar), treat as rank-1: axis 0 and -1 are both valid.
+    return (axis == 0) || (axis == -1 && rank == 0) || (-rank <= axis && axis < rank);
```
Comment (praasz) [inline src/core/src/validation_util.cpp]:
```suggestion
    return (axis == 0) || (axis == -1 ) || (-rank <= axis && axis < rank);
```
We can assume that axis 0, -1 is valid for any rank.

Or there are some issue in tests?

File: src/core/src/validation_util.cpp [hunk 6]
```diff
@@ -333,6 +334,9 @@ void validate_axis(const int64_t axis, const Rank& rank, const Node& node) {
 }
 
 size_t normalize_axis(const int64_t axis, const int64_t rank) {
+    // For rank-0 (scalar), any valid axis maps to 0.
+    if (rank == 0)
+        return 0;
     return static_cast<size_t>(normalize(axis, rank));
```
Comment (praasz) [inline src/core/src/validation_util.cpp]:
```suggestion
    return rank > 0 ? static_cast<size_t>(normalize(axis, rank)) : 0U;
```

Instead add `// For rank-0 (scalar), any valid axis maps to 0.` just update doxy of this function

File: src/core/src/validation_util.cpp
Comment (praasz) [inline src/core/src/validation_util.cpp]: Update core test in `validation_utils.cpp` file for this changes

File: src/core/tests/type_prop/log_softmax.cpp [hunk 7]
```diff
@@ -23,7 +23,7 @@ TEST(type_prop, log_softmax_incorrect_axis) {
         auto log_softmax_func = make_shared<op::v5::LogSoftmax>(data, 3);
         FAIL() << "LogSoftmax node was created with incorrect axis.";
     } catch (const NodeValidationFailure& error) {
-        EXPECT_HAS_SUBSTRING(error.what(), std::string("Reduction axis (3) is out of bounds"));
+        EXPECT_HAS_SUBSTRING(error.what(), std::string("Axis 3 out of the tensor rank range"));
     }
```
Comment (praasz) [inline src/core/tests/type_prop/log_softmax.cpp]:
```suggestion
    OV_EXPECT_THROW(std::ignore =  make_shared<op::v5::LogSoftmax>(data, 3),
                    NodeValidationFailure,
                    HasSubstr("Axis 3 out of the tensor rank range"));
```
Optional: the try catch could be replaced by assertion macro and checking exception message is optional, make like some sub-string or even skip. the exception is most important here.

File: src/core/tests/validation_utils.cpp [hunk 8]
```diff
@@ -248,14 +249,18 @@ TEST_P(NormalizeAxisTestP, is_axis_valid) {
 TEST_P(NormalizeAxisTestP, is_axis_valid_scalar_rank) {
     const auto& axis = std::get<0>(GetParam());
 
-    if (axis) {
+    if (axis == 0 || axis == -1) {
+        // axis=0 and axis=-1 are both valid for rank-0 (scalar treated as rank-1)
+        EXPECT_TRUE(util::is_axis_valid(axis, 0));
+    } else if (axis) {
         EXPECT_FALSE(util::is_axis_valid(axis, 0));
     }
```
Comment (praasz) [inline src/core/tests/validation_utils.cpp:256]:
```suggestion
    if (axis > 0 {
        EXPECT_FALSE(util::is_axis_valid(axis, 0));
    } else {
        EXPECT_TRUE(util::is_axis_valid(axis, 0));
    }
```

File: src/common/transformations/tests/op_conversions/softmax_decomposition_test.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+#include <memory>
+#include <string>
+
+#include "common_test_utils/ov_test_utils.hpp"
+#include "openvino/core/model.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/softmax.hpp"
+#include "openvino/pass/manager.hpp"
+using namespace ov;
+using namespace testing;
```
Comment (praasz) [inline src/common/transformations/tests/op_conversions/softmax_decomposition_test.cpp]: Remove and wrap tests into 
namespace ov::test{

File: src/core/src/op/softmax.cpp [hunk 10]
```diff
@@ -136,15 +131,25 @@ bool Softmax::evaluate(TensorVector& outputs, const TensorVector& inputs) const
...
-                    "Reduction axis (",
-                    m_axis,
-                    ") is out of bounds (argument shape: ",
-                    input_shape,
-                    ").");
-    const auto axis = static_cast<size_t>(ov::util::normalize(m_axis, rank));
-
     outputs[0].set_shape(input_shape);
+
+    // Rank-0 (scalar): treat as 1-element 1D tensor; softmax of a single value is always 1.
```
Comment (praasz) [inline src/core/src/op/softmax.cpp]: There is no need to put this comment every where

File: src/core/src/op/softmax.cpp [hunk 11]
```diff
@@ -136,15 +131,25 @@ bool Softmax::evaluate(TensorVector& outputs, const TensorVector& inputs) const
...
     const auto rank = static_cast<int64_t>(input_shape.size());
-    OPENVINO_ASSERT(-rank <= m_axis && m_axis < rank,
-                    "Reduction axis (",
-                    m_axis,
-                    ") is out of bounds (argument shape: ",
-                    input_shape,
-                    ").");
-    const auto axis = static_cast<size_t>(ov::util::normalize(m_axis, rank));
-
     outputs[0].set_shape(input_shape);
```
Comment (praasz) [inline src/core/src/op/softmax.cpp:138]: Should this `eval_shape` be set here?

File: src/core/src/op/softmax.cpp [hunk 12]
```diff
@@ -136,15 +131,25 @@ bool Softmax::evaluate(TensorVector& outputs, const TensorVector& inputs) const
...
+    } else {
+        eval_shape = input_shape;
+        OPENVINO_ASSERT(ov::util::is_axis_valid(m_axis, rank),
+                        "Reduction axis (",
+                        m_axis,
+                        ") is out of bounds (argument shape: ",
+                        input_shape,
+                        ").");
+        axis = ov::util::normalize_axis(m_axis, rank);
+    }
```
Comment (praasz) [inline src/core/src/op/softmax.cpp]:
```suggestion
    const auto& eval_shape = (rank == 0) ? Shape{1} : input_shape;
    const auto axis = (rank == 0) ? 0 : try_normalize_axis(m_axis, rank, *this);
```
Usually if there is same check in consecutive lines compiler well optimize in term of bin size.

File: src/core/src/op/softmax.cpp [hunk 13]
```diff
@@ -42,14 +42,16 @@ bool Softmax::visit_attributes(AttributeVisitor& visitor) {
...
+    if (const auto& rank = input_shape.rank(); rank.is_static()) {
+        const auto r = static_cast<size_t>(rank.get_length());
         NODE_VALIDATION_CHECK(this,
-                              m_axis < static_cast<size_t>(input_shape.rank().get_length()),
+                              (r == 0) ? (m_axis == 0) : (m_axis < r),
                               "Reduction axis (",
                               m_axis,
                               ") is out of bounds (argument shape: ",
                               input_shape,
                               ").");
```
Comment (praasz) [inline src/core/src/op/softmax.cpp:53]: Update validation util should also do check correctly can it be replaced?

File: src/core/tests/validation_utils.cpp (hunk 8 above)
Comment (evkotov) [reply src/core/tests/validation_utils.cpp:256]: Simplified the test (removed comment and else if (axis) guard), but used axis == 0 || axis == -1 instead of axis > 0. The parametrized suite includes negative values like -5 and -3, which are invalid for rank-0 but axis > 0 would classify them as valid (since -5 > 0 is false → EXPECT_TRUE(is_axis_valid(-5, 0)) → fail).

File: src/core/src/op/softmax.cpp (hunk 13 above)
Comment (evkotov) [reply src/core/src/op/softmax.cpp:53]: v1::Softmax stores axis as size_t (unsigned). When a user passes -1, it becomes (size_t)-1 = 18446744073709551615. If we call validate_axis(static_cast<int64_t>(m_axis), rank, *this), the cast converts that huge value back to -1, and is_axis_valid(-1, rank) now returns true for any rank. This would break the existing softmax_out_of_bound_axis test which expects v1::Softmax to reject negative axes.

**Review by praasz — APPROVED** [review]

File: src/frontends/pytorch/src/op/log_softmax.cpp [hunk 14]
```diff
@@ -36,6 +38,20 @@ OutputVector translate_log_softmax_common(const NodeContext& context, bool is_fx
...
+    const auto& input_pshape = input.get_partial_shape();
+    if (input_pshape.rank().is_static() && input_pshape.rank().get_length() == 0) {
+        FRONT_END_OP_CONVERSION_CHECK(dim == 0 || dim == -1,
+                                      "LogSoftmax on scalar input only supports dim=0 or dim=-1, got dim=",
+                                      dim);
+        auto zero_const = v0::Constant::create(element::i32, Shape{}, {0});
+        input = context.mark_node(std::make_shared<v0::Unsqueeze>(input, zero_const));
+        auto log_softmax = context.mark_node(std::make_shared<v5::LogSoftmax>(input, 0));
+        auto result = context.mark_node(std::make_shared<v0::Squeeze>(log_softmax, zero_const));
+        return {result};
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/log_softmax.cpp]: Why do we need this workaround if you changed core op to support 0-rank?

File: src/frontends/pytorch/src/op/softmax.cpp [hunk 15]
```diff
@@ -21,6 +23,20 @@ OutputVector translate_softmax_common(const NodeContext& context, const bool con
...
+    const auto& x_pshape = x.get_partial_shape();
+    if (x_pshape.rank().is_static() && x_pshape.rank().get_length() == 0) {
+        FRONT_END_OP_CONVERSION_CHECK(axis == 0 || axis == -1,
+                                      "Softmax on scalar input only supports dim=0 or dim=-1, got dim=",
+                                      axis);
+        auto zero_const = v0::Constant::create(element::i32, Shape{}, {0});
+        x = context.mark_node(std::make_shared<v0::Unsqueeze>(x, zero_const));
+        auto softmax = context.mark_node(std::make_shared<v8::Softmax>(x, int64_t{0}));
+        auto result = context.mark_node(std::make_shared<v0::Squeeze>(softmax, zero_const));
+        return {result};
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/softmax.cpp]: Same here: Why do we need this workaround if you changed core op to support 0-rank?
